amy/emails/tests/actions/test_instructor_training_approaching_strategy.py

```diff
@@ -0,0 +1,222 @@
+        # Act
+        result = instructor_training_approaching_strategy(self.event)
+        # Assert
+        self.assertEqual(result, StrategyEnum.NOOP)
+
+
+class TestRunInstructorTrainingApproachingStrategy(TestCase):
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_signal",
+    )
+    def test_strategy_calls_create_signal(
+        self,
+        mock_instructor_training_approaching_signal,
+    ) -> None:
+        # Arrange
+        strategy = StrategyEnum.CREATE
+        request = RequestFactory().get("/")
+        event = Event(start=datetime.today())
+
+        # Act
+        run_instructor_training_approaching_strategy(strategy, request, event)
+
+        # Assert
+        mock_instructor_training_approaching_signal.send.assert_called_once_with(
+            sender=event,
+            request=request,
+            event=event,
+            event_start_date=event.start,
+        )
+
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_update_signal",
+    )
+    def test_strategy_calls_update_signal(
+        self,
+        mock_instructor_training_approaching_update_signal,
+    ) -> None:
+        # Arrange
+        strategy = StrategyEnum.UPDATE
+        request = RequestFactory().get("/")
+        event = Event(start=datetime.today())
+
+        # Act
+        run_instructor_training_approaching_strategy(strategy, request, event)
+
+        # Assert
+        mock_instructor_training_approaching_update_signal.send.assert_called_once_with(
+            sender=event,
+            request=request,
+            event=event,
+            event_start_date=event.start,
+        )
+
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_remove_signal",
+    )
+    def test_strategy_calls_remove_signal(
+        self,
+        mock_instructor_training_approaching_remove_signal,
+    ) -> None:
+        # Arrange
+        strategy = StrategyEnum.REMOVE
+        request = RequestFactory().get("/")
+        event = Event(start=datetime.today())
+
+        # Act
+        run_instructor_training_approaching_strategy(strategy, request, event)
+
+        # Assert
+        mock_instructor_training_approaching_remove_signal.send.assert_called_once_with(
+            sender=event,
+            request=request,
+            event=event,
+            event_start_date=event.start,
+        )
+
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_signal",
+    )
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_update_signal",
+    )
+    @patch(
+        "emails.actions.instructor_training_approaching."
+        "instructor_training_approaching_remove_signal",
+    )
+    def test_invalid_strategy_no_signal_called(
```

Comment:
IMO invalid strategy selection is different to a deliberate noop selection - there could be tests for both those cases.

Reply:
Expanded with the changes for more exhaustive mapping used in the strategy.